mysql/access_rules_test.go

```diff
@@ -0,0 +1,175 @@
+
+	sClient, aClient, err := getAccessRulesTestClients()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	sInstance, err := sClient.createTestServiceInstance()
+	if err != nil {
+		t.Fatalf("Error creating Service Instance: %s", err)
+	}
+
+	instanceName := sInstance.ServiceName
+	defer destroyServiceInstance(t, sClient, instanceName)
+
+	createAccessRuleInput := &CreateAccessRuleInput{
+		ServiceInstanceID: instanceName,
+		Description:       _Service_AccessRule_Description,
+		Destination:       _Service_AccessRule_Destination,
+		Ports:             _Service_AccessRule_Ports,
+		Protocol:          _Service_AccessRule_Protocol,
+		RuleName:          _Service_AccessRule_Name,
+		Source:            _Service_AccessRule_Source,
+		Status:            _Service_AccessRule_Status,
+	}
+
+	expected := &AccessRuleInfo{
+		Description: _Service_AccessRule_Description,
+		Destination: _Service_AccessRule_Destination,
+		Ports:       _Service_AccessRule_Ports,
+		Protocol:    _Service_AccessRule_Protocol,
+		RuleName:    _Service_AccessRule_Name,
+		Source:      _Service_AccessRule_Source,
+		Status:      _Service_AccessRule_Status,
+		RuleType:    "USER",
+	}
+
+	if err := aClient.CreateAccessRule(createAccessRuleInput); err != nil {
+		t.Fatalf("Error creating AccessRule: %s", err)
+	}
+
+	// Get Access Rule (Create only returns AccessRule name)
+	getInput := &GetAccessRuleInput{
+		ServiceInstanceID: instanceName,
+	}
+
+	allRulesResult, err := aClient.GetAllAccessRules(getInput)
+	if err != nil {
+		t.Fatalf("Error reading ALL AccessRules : %s", err)
+	}
+
+	if len(allRulesResult.AccessRules) == 0 {
+		t.Fatalf("Error reading ALL accessRules: Expected at least 1 rule. Got %d", len(allRulesResult.AccessRules))
+	}
+
+	// Read Result
+	getInput.Name = _Service_AccessRule_Name
+	ruleResult, err := aClient.GetAccessRule(getInput)
+
+	if err != nil {
+		t.Fatalf("Error reading AccessRule: %s", err)
+	}
+
+	// Test Assertions
+	if diff := pretty.Compare(ruleResult, expected); diff != "" {
+		t.Fatalf("Diff creating AccessRule: (-got, +want):\n%s", diff)
+	}
+
+	// Update Access Rule
+	updateInput := &UpdateAccessRuleInput{
+		ServiceInstanceID: instanceName,
+		Name:              _Service_AccessRule_Name,
+		Status:            "disabled",
+	}
+
+	if _, err := aClient.UpdateAccessRule(updateInput); err != nil {
+		t.Fatalf("Error updating AccessRule: %s", err)
+	}
+
+	// Re-Read Result
+	ruleResult, err = aClient.GetAccessRule(getInput)
+	if err != nil {
+		t.Fatalf("Error reading AccessRule: %s", err)
+	}
+
+	// Change expected to match
+	expected.Status = "disabled"
+
+	// Test Assertions
+	if diff := pretty.Compare(ruleResult, expected); diff != "" {
+		t.Fatalf("Diff creating AccessRule: (-got, +want):\n%s", diff)
+	}
+
+	// We don't explicitly test the delete, because its called
+	// using the defer.
+
+	// Sleep for 3 minutes to prevent failure because the resource is still
+	// locked.
+
+	destroyAccessRule(t, aClient, instanceName, _Service_AccessRule_Name)
```

Comment:
We can defer this right after we create the access rule.

Reply:
I've tried that, but the delete fails. The REST API returns a error 400, and the whole test stops. Not too sure whats the cause of it either.

I added a comment on the failure into the test code, and left it there until i can figure out why its failing.

Reply:
I managed to fix this. Now defer works properly. The problem is linked to the WaitForAccessRuleDelete.